firewood/src/merkle/proof.rs

```diff
@@ -258,261 +259,150 @@ impl<N: AsRef<[u8]> + Send> Proof<N> {
 
-                if !key.contains_other(cur_key) {
-                    return Ok((addr, None, 0));
-                }
+                NodeType::Branch(n) => {
+                    if let Some(index) = key_nibbles.next() {
+                        let subproof = n
+                            .chd_encode()
+                            .get(index as usize)
+                            .and_then(|inner| inner.as_ref())
+                            .map(|data| &**data);
 
-                let subproof = SubProof::Data(n.data().to_vec());
+                        child_index = index as usize;
 
-                Ok((addr, subproof.into(), cur_key.len()))
-            }
-            NodeType::Extension(n) => {
-                let cur_key = n.path.0.as_ref();
+                        subproof
+                    } else {
+                        break n.value.as_ref().map(|data| data.to_vec());
+                    }
+                }
+            };
 
-                if !key.contains_other(cur_key) {
-                    return Ok((addr, None, 0));
+            match encoded_sub_proof {
+                None => break None,
+                Some(encoded) => {
+                    let hash = generate_subproof_hash(encoded)?;
+                    child_hash = hash;
                 }
+            }
 
-                let encoded = n.chd_encoded().ok_or(ProofError::InvalidData)?.to_vec();
-                let subproof = generate_subproof(encoded).map(Some)?;
+            parent_node_ref = child_node;
+        };
 
-                Ok((addr, subproof, cur_key.len()))
-            }
-            // If the node is the last one to be decoded, then no subproof to be extracted.
-            NodeType::Branch(_) if end_node => Ok((addr, None, 1)),
-            NodeType::Branch(_) if key.is_empty() => Err(ProofError::NoSuchNode),
-            NodeType::Branch(n) => {
-                // Check if the subproof with the given key exist.
-                #[allow(clippy::indexing_slicing)]
-                let index = key[0] as usize;
-                #[allow(clippy::indexing_slicing)]
-                let Some(data) = &n.chd_encode()[index] else {
-                    return Ok((addr, None, 1));
-                };
-                let subproof = generate_subproof(data.to_vec())?;
-                Ok((addr, Some(subproof), 1))
-            }
+        match sub_proof {
+            Some(data) => Ok(Some(data)),
+            None if allow_non_existent_node => Ok(None),
+            None => Err(ProofError::NodeNotInTrie),
         }
     }
 }
 
+fn decode_subproof<'a, S: ShaleStore<Node>, T, N: AsRef<[u8]>>(
+    merkle: &'a Merkle<S, T>,
+    proofs_map: &HashMap<[u8; 32], N>,
+    child_hash: &[u8; 32],
+) -> Result<ObjRef<'a>, ProofError> {
+    let child_proof = proofs_map
+        .get(child_hash)
+        .ok_or(ProofError::ProofNodeMissing)?;
+    let child_node = NodeType::decode(child_proof.as_ref())?;
+    merkle
+        .put_node(Node::from(child_node))
```

Comment:
This returns `MerkleError` and `ProofError` seems to already have a derived `into()` method from this type, so I think you can just use ? here. If that's wrong, just ignore this comment.